### DISCUSSION TITLE
OAuth2 Backend: How to get access for adding a binding to a Topic Exchange

### DISCUSSION THREAD

**Discussion by OLibutzki** [rabbitmq/rabbitmq-server#7145] [Questions] [answered]
Hello everyone,

I would like to a a binding from a queue to a topic exchange for a certain routing key using the Management UI.

My user is authenticated with the OAuth2 Backend and has the following client scopes:

- openid
- profile
- rabbitmq.*
- rabitmag.tag.administrator

The resource server id is set to rabbitmq:
`auth_oauth2.resource_server_id = rabbitmq`

With these permissions I am able to create queue, exchanges, vhosts,...
Unfortunately I cannot bind a queue to a topic exchange. The RabbitMQ server logs stats:
`[warning] <0.1018.0> Binding error: access to topic 'T' in exchange 'X' in vhost '/' refused for user 'U'`

After reading the docs carefully I understand that it seems to be an issue with Topic authorization, but I don't have a clue how to authorize the user. Any suggestions?

Thanks a lot.

--- EDIT ---
In the MQTT section of the [rabbitmq-oauth2-tutorial](https://github.com/rabbitmq/rabbitmq-oauth2-tutorial) I found this chapter:
> IMPORTANT: If you try to access the management ui and authenticate with UAA using rabbit_admin you
wont be able to do bind a queue with routing_key `test` to the `amq.topic` exchange because that user
in UAA does not have the required permissions. In our handcrafted token, you have granted ourselves the right permissions/scopes.

Sounds like the problem I have, but I have no idea how to solve this. I do not know what the handcrafted token is and who it grants the required permissions.

**Comment by michaelklishin** [comment, marked as answer]
Support for Auth 2 in AMQP 1.0 has just recently landed, MQTT is likely to be investigated next.

@MarcialRosales FYI.

**Reply by OLibutzki** [reply]
Thanks for responding. To clarify, I want to use AMQP, not MQTT.

I only referred to the MQTT part in the docs as I experienced the same behaviour while working with AMQP.

**Reply by MarcialRosales** [reply]
Hi @OLibutzki , unfortunately when you operate (i.e. `bind`, `unbind`, `publish`) on a topic exchange you have to explicitly grant access to the *routing-key*.  The exact format of the scope/permission is explained [here](https://www.rabbitmq.com/oauth2.html#scope-translation).  

In your case, you would need to grant the scopes `rabbitmq.read:*/*/* rabbitmq.write:*/*/*  rabbitmq.configure:*/*/* ` i..e the 3 operations on any vhost, on any resource (exchange or queue), and on any routing-key. 
 
With regards the hardcrafted token within the [MQTT section](https://github.com/rabbitmq/rabbitmq-oauth2-tutorial#mqtt-protocol) in the Oauth2 tutorial, it is shown within the section itself. I have copied it below though. See that it is granting three scopes, one for each operation with full access on any vhost, resource and routing-key  :   
```
{
  "scope": [
    "rabbitmq.write:*/*/*",
    "rabbitmq.configure:*/*/*",
    "rabbitmq.read:*/*/*"

  ],
  "extra_scope": "rabbitmq.tag:management",
  "aud": [
    "rabbitmq"
  ]
}
```

**Reply by OLibutzki** [reply]
Ok, thanks. Is it correct that variable expansion like `{username}` is not supported by the OAuth2 topic authorization?

**Reply by ansd** [reply]
Correct. As the [docs](https://www.rabbitmq.com/access-control.html#topic-authorisation) state:
>Internal (default) authorisation backend supports variable expansion in permission patterns. 

OAuth2 backend does not.

**Reply by OLibutzki** [reply]
Is it worth opening an issue or is it intended not to be supported?

**Reply by ansd** [reply]
I don't see a reason why variable expansion in topic permission patterns should only be supported by `rabbit_auth_backend_internal`. It makes sense to me to support it also for other auth backends. Feel free to open an issue and to submit a PR.

**Reply by OLibutzki** [reply]
Ok, I created an issue: #7178.

**Reply by brunomedeirosdedalus** [reply]
Did you try adding the permission `<resource_id>.configure:*/*/*, <resource_id>.read:*/*/* and <resource_id>.write:*/*/* `for instance?

**Reply by OLibutzki** [reply]
Yes, it works using this. My fault was that I just granted `<resource_id>.configure:*/*, <resource_id>.read:*/* and <resource_id>.write:*/* `. I have been missing the last asterisk for the topic auth part.
